internal/cron/pattern.go

```diff
@@ -0,0 +1,161 @@
+package cron
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/TBD54566975/ftl/internal/slices"
+	"github.com/alecthomas/participle/v2"
+	"github.com/alecthomas/participle/v2/lexer"
+)
+
+var (
+	lex = lexer.MustSimple([]lexer.SimpleRule{
+		{Name: "Whitespace", Pattern: `\s+`},
+		{Name: "Ident", Pattern: `\b[a-zA-Z_][a-zA-Z0-9_]*\b`},
+		{Name: "Comment", Pattern: `//.*`},
+		{Name: "String", Pattern: `"(?:\\.|[^"])*"`},
+		{Name: "Number", Pattern: `[0-9]+(?:\.[0-9]+)?`},
+		{Name: "Punct", Pattern: `[%/\-\_:[\]{}<>()*+?.,\\^$|#~!\'@]`},
+	})
+
+	parserOptions = []participle.Option{
+		participle.Lexer(lex),
+		participle.Elide("Whitespace"),
+		participle.Unquote(),
+		participle.Map(func(token lexer.Token) (lexer.Token, error) {
+			token.Value = strings.TrimSpace(strings.TrimPrefix(token.Value, "//"))
+			return token, nil
+		}, "Comment"),
+	}
+
+	parser = participle.MustBuild[Pattern](parserOptions...)
+)
+
+type Pattern struct {
+	Components []Component `parser:"@@ @@ @@ @@ @@ @@? @@?"`
+}
+
+func (p Pattern) String() string {
+	return strings.Join(slices.Map(p.Components, func(component Component) string {
+		return component.String()
+	}), " ")
+}
+
+func (p Pattern) standardizedComponents() ([]Component, error) {
+	switch len(p.Components) {
+	case 5:
+		// Convert "a b c d e" -> "0 a b c d e *"
+		components := make([]Component, 7)
+		components[0] = newComponentWithValue(0)
+		copy(components[1:], p.Components)
+		components[6] = newComponentWithFullRange()
+		return components, nil
+	case 6:
+		// Might be two different formats unfortunately.
```

Comment:
Are these semantics the same as other implementations?

Reply:
I haven't found a neat single place that defines it well, but this part of wikipedia helps show its a bit of a wild west https://en.wikipedia.org/wiki/Cron#Cron_expression
Note the year being optional (and not in standard versions) and that "In some uses of the cron format there is also a seconds field at the beginning of the pattern. In that case, the cron expression is a string comprising 6 or 7 fields."

I looked into how gronx handled this and they look for a 4 digit number to determine if it is a year, which is what I've done here as well.
https://github.com/adhocore/gronx/blob/d84c38971549dda6eb2803d83756e340fd40c297/gronx.go#L105